File: src/veil/mcp_server/tools.py
"""MCP tool contract (SPEC.md §6, §13, §18.3).

The schemas here are the structural security guarantee: there is no property
capable of carrying credential content, every object is closed
(``additionalProperties: false``), and arguments are screened for both
secret-shaped field names and credential-shaped values before they reach the
broker.
"""

from __future__ import annotations

import json
import re
from collections.abc import Mapping
from typing import Any

from veil.adapters.registry import AdapterRegistry
from veil.broker import SecretBroker, parse_store_params
from veil.errors import ErrorCode, PublicError, VeilError, veil_error
from veil.logging_ import AuditLogger
from veil.model import Environment, WriteMode
from veil.redaction import looks_like_credential, safe_display

STORE_TOOL = "secret.store"
STATUS_TOOL = "secret.status"
CANCEL_TOOL = "secret.cancel"
REVISE_TOOL = "secret.revise"
DESTINATIONS_TOOL = "secret.destinations"

#: Field names that would betray an attempt to smuggle credential material
#: through an unmodelled property (SPEC.md §6, §26.5).
SECRET_SHAPED_FIELD = re.compile(
    r"(?:^|[_.\-])?(secret[_-]?value|secretvalue|raw[_-]?secret|rawsecret|password|passwd|pwd|"
    r"passphrase|credential|credentials|api[_-]?key[_-]?value|token[_-]?value|private[_-]?key|"
    r"client[_-]?secret|access[_-]?key|content|payload|value|data|blob|bytes|material)"
    r"(?:$|[_.\-])?",
    re.IGNORECASE,
)

MAX_ARGUMENT_DEPTH = 6
#: A blocking status call occupies the single-threaded stdio loop, so it is kept
#: short: the human's window runs on its own thread and is never blocked by it.
MAX_STATUS_WAIT_SECONDS = 60


class ToolResult:
    __slots__ = ("is_error", "payload")

    def __init__(self, payload: dict[str, Any], *, is_error: bool = False) -> None:
        self.payload = payload
        self.is_error = is_error

    def to_mcp(self) -> dict[str, Any]:
        text = json.dumps(self.payload, indent=2, sort_keys=True, ensure_ascii=False)
        return {
            "content": [{"type": "text", "text": text}],
            "structuredContent": self.payload,
            "isError": self.is_error,
        }


class ToolRouter:
    def __init__(
        self,
        broker: SecretBroker,
        registry: AdapterRegistry,
        logger: AuditLogger,
    ) -> None:
        self.broker = broker
        self.registry = registry
        self.log = logger

    # -- schema ------------------------------------------------------------

    def list_tools(self) -> list[dict[str, Any]]:
        return [
            {
                "name": STORE_TOOL,
                "title": "Request that the user store a credential",
                "description": (
                    "Ask the human to provide a credential and have Veil write it to the "
                    "destination described here. The credential value is never passed through "
                    "this tool, never returned by it, and never becomes visible to the model: "
                    "the user enters it in Veil's own trusted window. Share the returned "
                    "authorization_url with the user, then poll secret.status."
                ),
                "inputSchema": self._store_schema(),
                "annotations": {"destructiveHint": True, "openWorldHint": True},
            },
            {
                "name": STATUS_TOOL,
                "title": "Check a credential request",
                "description": (
                    "Return the non-sensitive status of a credential request. Never returns "
                    "credential material."
                ),
                "inputSchema": {
                    "type": "object",
                    "additionalProperties": False,
                    "required": ["request_id"],
                    "properties": {
                        "request_id": {"type": "string", "maxLength": 64},
                        "wait_seconds": {
                            "type": "integer",
                            "minimum": 0,
                            "maximum": 60,
                            "description": "Optionally block until the request reaches a "
                            "terminal state or this many seconds elapse.",
                        },
                    },
                },
                "annotations": {"readOnlyHint": True},
            },
            {
                "name": CANCEL_TOOL,
                "title": "Cancel a credential request",
                "description": "Cancel a pending request. Any credential already entered is "
                "destroyed.",
                "inputSchema": {
                    "type": "object",
                    "additionalProperties": False,
                    "required": ["request_id"],
                    "properties": {
                        "request_id": {"type": "string", "maxLength": 64},
                        "reason": {"type": "string", "maxLength": 200},
                    },
                },
            },
            {
                "name": REVISE_TOOL,
                "title": "Replace a credential request with a corrected one",
                "description": (
                    "Cancel a pending request and create a new one. The original authorization "
                    "is invalidated and the human must authorize the new operation from "
                    "scratch; an authorized operation can never be edited in place."
                ),
                "inputSchema": self._revise_schema(),
            },
            {
                "name": DESTINATIONS_TOOL,
                "title": "List available destinations",
                "description": "List the destinations this Veil instance can write to, with "
                "the target fields each one expects.",
                "inputSchema": {"type": "object", "additionalProperties": False, "properties": {}},
                "annotations": {"readOnlyHint": True},
            },
        ]

    def _store_schema(self) -> dict[str, Any]:
        target_properties: dict[str, Any] = {}
        for adapter in self.registry:
            schema = adapter.target_schema()
            for key, definition in schema.get("properties", {}).items():
                target_properties.setdefault(key, definition)
        return {
            "type": "object",
            "additionalProperties": False,
            "required": ["destination", "name", "target"],
            "properties": {
                "destination": {
                    "type": "string",
                    "enum": list(self.registry.ids()),
                    "description": "Which destination adapter should receive the credential.",
                },
                "name": {
                    "type": "string",
                    "maxLength": 128,
                    "description": (
                        "Logical name of the credential, e.g. STRIPE_SECRET_KEY. This is a "
                        "label, never the credential value."
                    ),
                },
                "target": {
                    "type": "object",
                    "additionalProperties": False,
                    "properties": target_properties,
                    "description": "Where the credential goes. Fields depend on the "
                    "destination; call secret.destinations for the exact contract.",
                },
                "write_mode": {
                    "type": "string",
                    "enum": [str(mode) for mode in WriteMode],
                    "default": str(WriteMode.CREATE),
                },
                "environment": {
                    "type": "string",
                    "enum": [str(env) for env in Environment],
                    "description": (
                        "Environment you believe this destination belongs to. Advisory only: "
                        "Veil classifies the destination itself and uses the stricter of the "
                        "two."
                    ),
                },
                "description": {
                    "type": "string",
                    "maxLength": 512,
                    "description": "Short human-readable purpose, shown to the user.",
                },
            },
        }

    def _revise_schema(self) -> dict[str, Any]:
        schema = self._store_schema()
        schema["properties"] = {
            "request_id": {"type": "string", "maxLength": 64},
            **schema["properties"],
        }
        schema["required"] = ["request_id", *schema["required"]]
        return schema

    # -- dispatch ----------------------------------------------------------

    def call(self, name: object, arguments: object) -> ToolResult:
        args: Mapping[str, Any]
        if arguments is None:
            args = {}
        elif isinstance(arguments, Mapping):
            args = arguments
        else:
            return _error(veil_error(ErrorCode.INVALID_ARGUMENTS, "Arguments must be an object."))

        try:
            self._screen_arguments(name, args)
            if name == STORE_TOOL:
                return self._store(args)
            if name == STATUS_TOOL:
                return self._status(args)
            if name == CANCEL_TOOL:
                return self._cancel(args)
            if name == REVISE_TOOL:
                return self._revise(args)
            if name == DESTINATIONS_TOOL:
                return self._destinations()
            raise veil_error(ErrorCode.INVALID_ARGUMENTS, "Unknown tool.")
        except VeilError as exc:
            return _error(exc)
        except Exception:
            self.log.error("tool_call_failed", tool=safe_display(str(name), max_length=40))
            return ToolResult(
                PublicError(
                    ErrorCode.INTERNAL_ERROR,
                    "The request could not be processed.",
                ).to_dict(),
                is_error=True,
            )

    # -- screening ---------------------------------------------------------

    def _screen_arguments(self, tool: object, args: Mapping[str, Any]) -> None:
        """Reject covert secret transport before anything else looks at the args."""

        allowed_keys = {
            "destination",
            "name",
            "target",
            "write_mode",
            "environment",
            "description",
            "request_id",
            "reason",
            "wait_seconds",
        } | {"project", "secret", "collection", "document", "field", "path", "key"}

        findings: list[str] = []

        def walk(node: Any, depth: int) -> None:
            if depth > MAX_ARGUMENT_DEPTH:
                findings.append("nesting")
                return
            if isinstance(node, Mapping):
                for key, value in node.items():
                    key_text = str(key)
                    if key_text not in allowed_keys and SECRET_SHAPED_FIELD.search(key_text):
                        findings.append(f"field:{safe_display(key_text, max_length=40)}")
                    walk(value, depth + 1)
            elif isinstance(node, list | tuple):
                for item in node:
                    walk(item, depth + 1)
            elif isinstance(node, str) and looks_like_credential(node):
                findings.append("credential-shaped-value")

        walk(args, 0)
        if findings:
            self.log.security(
                "secret_shaped_argument_rejected",
                tool=safe_display(str(tool), max_length=40),
                detail={"findings": sorted(set(findings))[:5]},
            )
            raise veil_error(
                ErrorCode.FORBIDDEN_FIELD,
                "This tool never accepts credential material. Remove the offending field and "
                "let the user enter the value in Veil.",
            )

    # -- handlers ----------------------------------------------------------

    def _store(self, args: Mapping[str, Any]) -> ToolResult:
        params = parse_store_params(args, self.registry)
        request = self.broker.create_request(params)
        return ToolResult(self._pending_payload(request.request_id))

    def _revise(self, args: Mapping[str, Any]) -> ToolResult:
        request_id = args.get("request_id")
        if not isinstance(request_id, str):
            raise veil_error(ErrorCode.INVALID_ARGUMENTS, "A request_id is required.")
        rest = {k: v for k, v in args.items() if k != "request_id"}
        params = parse_store_params(rest, self.registry)
        request = self.broker.revise(request_id, params)
        payload = self._pending_payload(request.request_id)
        payload["previous_request_id"] = safe_display(request_id, max_length=64)
        payload["notice"] = (
            "The previous authorization was invalidated. The user must authorize this "
            "operation again."
        )
        return ToolResult(payload)

    def _status(self, args: Mapping[str, Any]) -> ToolResult:
        request_id = args.get("request_id")
        wait = args.get("wait_seconds", 0)
        if isinstance(wait, bool) or not isinstance(wait, int) or wait < 0:
            wait = 0
        wait = min(wait, MAX_STATUS_WAIT_SECONDS)
        if wait:
            return ToolResult(self.broker.wait_for_terminal(request_id, float(wait)))
        return ToolResult(self.broker.public_status(request_id))

    def _cancel(self, args: Mapping[str, Any]) -> ToolResult:
        request_id = args.get("request_id")
        reason = args.get("reason")
        self.broker.cancel(
            request_id,
            reason=safe_display(str(reason), max_length=80) if reason else "agent_cancelled",
        )
        return ToolResult(self.broker.public_status(request_id))

    def _destinations(self) -> ToolResult:
        return ToolResult(
            {
                "destinations": [
                    {
                        "id": adapter.id,
                        "title": adapter.display_name,
                        "destination_class": str(adapter.destination_class),
                        "write_modes": [str(m) for m in adapter.supported_write_modes()],
                        "target_schema": adapter.target_schema(),
                    }
                    for adapter in self.registry
                ],
                "note": (
                    "No destination accepts a credential value through this API. Veil collects "
                    "the value from the user directly."
                ),
            }
        )

    def _pending_payload(self, request_id: str) -> dict[str, Any]:
        status = self.broker.public_status(request_id)
        status["next_step"] = (
            "Show the authorization_url to the user. They will review the destination and "
            "enter the credential in Veil's own window; you will not see the value."
        )
        return status


def _error(exc: VeilError) -> ToolResult:
    return ToolResult(exc.public.to_dict(), is_error=True)

File: src/mcp/tools.ts
/**
 * MCP tool contract (SPEC.md §6, §13, §18.3).
 *
 * The schemas here are the structural security guarantee: there is no property
 * capable of carrying credential content, every object is closed
 * (`additionalProperties: false`), and arguments are screened for both
 * secret-shaped field names and credential-shaped values before they reach the
 * broker.
 *
 * Descriptions follow the Anthropic connector review criteria: they state what
 * the tool does, what it returns and what it does not do, and they never
 * instruct the model how to behave. Workflow guidance belongs in the server's
 * `instructions`, not in a tool description.
 *
 * Why not elicitation? The spec-native way to collect user input mid-tool would
 * deliver the value back through the MCP protocol and into the model's context —
 * exactly what Veil exists to prevent (SPEC.md §5). The out-of-band browser
 * window is not a workaround for missing elicitation support; it is the point.
 */

import { z } from 'zod';

import type { AdapterRegistry } from '../adapters/registry.js';
import { parseStoreParams, type SecretBroker } from '../broker.js';
import { ErrorCode, VeilError, publicError, publicErrorToJSON, veilError } from '../errors.js';
import type { AuditLogger } from '../logging.js';
import { Environment, WriteMode } from '../model.js';
import { looksLikeCredential, safeDisplay } from '../redaction.js';

export const STORE_TOOL = 'secret_store';
export const STATUS_TOOL = 'secret_status';
export const CANCEL_TOOL = 'secret_cancel';
export const REVISE_TOOL = 'secret_revise';
export const DESTINATIONS_TOOL = 'secret_destinations';

/**
 * Field names that would betray an attempt to smuggle credential material
 * through an unmodelled property (SPEC.md §6, §26.5).
 */
export const SECRET_SHAPED_FIELD =
  /(secret[_-]?value|secretvalue|raw[_-]?secret|rawsecret|password|passwd|pwd|passphrase|credential|credentials|api[_-]?key[_-]?value|token[_-]?value|private[_-]?key|client[_-]?secret|access[_-]?key|content|payload|value|data|blob|bytes|material)/i;

export const MAX_ARGUMENT_DEPTH = 6;
/**
 * A blocking status call occupies the single-threaded stdio loop, so it is kept
 * short: the human's window runs independently and is never blocked by it.
 */
export const MAX_STATUS_WAIT_SECONDS = 60;

const ALLOWED_ARGUMENT_KEYS = new Set([
  'destination',
  'name',
  'target',
  'write_mode',
  'environment',
  'description',
  'request_id',
  'reason',
  'wait_seconds',
  'project',
  'secret',
  'collection',
  'document',
  'field',
  'path',
  'key',
]);

export interface McpTool {
  readonly name: string;
  readonly title: string;
  readonly description: string;
  readonly inputSchema: Record<string, unknown>;
  readonly annotations: Record<string, unknown>;
}

export interface ToolResult {
  readonly payload: Record<string, unknown>;
  readonly isError: boolean;
}

export function toolResultToMcp(result: ToolResult): Record<string, unknown> {
  return {
    content: [{ type: 'text', text: JSON.stringify(result.payload, null, 2) }],
    structuredContent: result.payload,
    isError: result.isError,
  };
}

export const statusArgsSchema = z
  .object({
    request_id: z.string().max(64),
    wait_seconds: z.number().int().min(0).max(MAX_STATUS_WAIT_SECONDS).optional(),
  })
  .strict();

export const cancelArgsSchema = z
  .object({ request_id: z.string().max(64), reason: z.string().max(200).optional() })
  .strict();

export class ToolRouter {
  constructor(
    readonly broker: SecretBroker,
    readonly registry: AdapterRegistry,
    readonly log: AuditLogger,
  ) {}

  listTools(): McpTool[] {
    return [
      {
        name: STORE_TOOL,
        title: 'Request credential storage',
        description:
          'Creates a request to store a credential at the described destination and opens ' +
          "Veil's own authorization window on the user's machine, where the user reviews " +
          'the destination and types the value. Returns a request id, the normalized ' +
          'destination, a risk classification and the request state — never the credential ' +
          'value, which no Veil tool accepts or returns. Does not write anything by itself: ' +
          'the write happens only after the user authorizes it. Use secret_status to follow ' +
          'the request, and secret_revise to change an already-authorized destination.',
        inputSchema: this.storeSchema(),
        annotations: {
          title: 'Request credential storage',
          readOnlyHint: false,
          destructiveHint: true,
          idempotentHint: false,
          openWorldHint: true,
        },
      },
      {
        name: STATUS_TOOL,
        title: 'Check a credential request',
        description:
          'Returns the current state of one credential request: its stage, destination, ' +
          'risk, and the outcome once it is finished. Never returns credential material. ' +
          'Reads only — use secret_store to create a request.',
        inputSchema: {
          type: 'object',
          additionalProperties: false,
          required: ['request_id'],
          properties: {
            request_id: {
              type: 'string',
              maxLength: 64,
              description: 'Identifier returned by secret_store or secret_revise.',
            },
            wait_seconds: {
              type: 'integer',
              minimum: 0,
              maximum: MAX_STATUS_WAIT_SECONDS,
              description:
                'Block until the request reaches a terminal state or this many seconds ' +
                'elapse. Omit or use 0 to return immediately.',
            },
          },
        },
        annotations: {
          title: 'Check a credential request',
          readOnlyHint: true,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: false,
        },
      },
      {
        name: CANCEL_TOOL,
        title: 'Cancel a credential request',
        description:
          'Cancels a pending credential request and destroys any value the user already ' +
          'entered. Returns the final state. Cannot cancel a request that is already ' +
          'writing to its destination, and cannot undo a completed write.',
        inputSchema: {
          type: 'object',
          additionalProperties: false,
          required: ['request_id'],
          properties: {
            request_id: {
              type: 'string',
              maxLength: 64,
              description: 'Identifier of the request to cancel.',
            },
            reason: {
              type: 'string',
              maxLength: 200,
              description: 'Short, non-sensitive note recorded in the audit log.',
            },
          },
        },
        annotations: {
          title: 'Cancel a credential request',
          readOnlyHint: false,
          destructiveHint: true,
          idempotentHint: true,
          openWorldHint: false,
        },
      },
      {
        name: REVISE_TOOL,
        title: 'Replace a credential request',
        description:
          'Cancels a pending request and creates a replacement with different details. The ' +
          'original authorization is invalidated and the user authorizes the new operation ' +
          'from scratch; an authorized operation is never edited in place. Returns the new ' +
          'request id. Use secret_cancel when no replacement is wanted.',
        inputSchema: this.reviseSchema(),
        annotations: {
          title: 'Replace a credential request',
          readOnlyHint: false,
          destructiveHint: true,
          idempotentHint: false,
          openWorldHint: true,
        },
      },
      {
        name: DESTINATIONS_TOOL,
        title: 'List credential destinations',
        description:
          'Lists the destinations this Veil instance can write to, with the target fields ' +
          'and write modes each one accepts. Reads only; creates no request.',
        inputSchema: { type: 'object', additionalProperties: false, properties: {} },
        annotations: {
          title: 'List credential destinations',
          readOnlyHint: true,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: false,
        },
      },
    ];
  }

  storeSchema(): Record<string, unknown> {
    const targetProperties: Record<string, unknown> = {};
    for (const adapter of this.registry.list()) {
      for (const [key, definition] of Object.entries(adapter.targetSchema().properties)) {
        targetProperties[key] ??= definition;
      }
    }
    return {
      type: 'object',
      additionalProperties: false,
      required: ['destination', 'name', 'target'],
      properties: {
        destination: {
          type: 'string',
          enum: [...this.registry.ids()],
          description: 'Destination adapter that will receive the credential.',
        },
        name: {
          type: 'string',
          maxLength: 128,
          description:
            'Logical name of the credential, e.g. STRIPE_SECRET_KEY. A label only: this ' +
            'field must never contain the credential value.',
        },
        target: {
          type: 'object',
          additionalProperties: false,
          properties: targetProperties,
          description:
            'Where the credential goes. Accepted fields depend on the destination; ' +
            'secret_destinations returns the exact contract for each.',
        },
        write_mode: {
          type: 'string',
          enum: Object.values(WriteMode),
          default: WriteMode.CREATE,
          description:
            "'create' fails if a value already exists, 'new-version' adds a version, " +
            "'replace' overwrites. Availability depends on the destination.",
        },
        environment: {
          type: 'string',
          enum: Object.values(Environment),
          description:
            'Environment this destination is believed to belong to. Advisory only: Veil ' +
            'classifies the destination itself and applies the stricter of the two.',
        },
        description: {
          type: 'string',
          maxLength: 512,
          description: 'Short, non-sensitive purpose shown to the user on the review screen.',
        },
      },
    };
  }

  reviseSchema(): Record<string, unknown> {
    const schema = this.storeSchema();
    const properties = schema.properties as Record<string, unknown>;
    return {
      ...schema,
      required: ['request_id', ...(schema.required as string[])],
      properties: {
        request_id: {
          type: 'string',
          maxLength: 64,
          description: 'Identifier of the request being replaced.',
        },
        ...properties,
      },
    };
  }

  async call(name: unknown, args: unknown): Promise<ToolResult> {
    const argumentObject: Record<string, unknown> =
      args === null || args === undefined
        ? {}
        : typeof args === 'object' && !Array.isArray(args)
          ? (args as Record<string, unknown>)
          : (undefined as never);

    if (argumentObject === undefined) {
      return errorResult(veilError(ErrorCode.INVALID_ARGUMENTS, 'Arguments must be an object.'));
    }

    try {
      this.#screenArguments(name, argumentObject);
      switch (name) {
        case STORE_TOOL:
          return await this.#store(argumentObject);
        case STATUS_TOOL:
          return await this.#status(argumentObject);
        case CANCEL_TOOL:
          return this.#cancel(argumentObject);
        case REVISE_TOOL:
          return await this.#revise(argumentObject);
        case DESTINATIONS_TOOL:
          return this.#destinations();
        default:
          throw veilError(ErrorCode.INVALID_ARGUMENTS, 'Unknown tool.');
      }
    } catch (error) {
      if (error instanceof VeilError) return errorResult(error);
      this.log.error('tool_call_failed', { tool: safeDisplay(String(name), 40) });
      return {
        payload: publicErrorToJSON(
          publicError(ErrorCode.INTERNAL_ERROR, 'The request could not be processed.'),
        ),
        isError: true,
      };
    }
  }

  /** Reject covert secret transport before anything else looks at the args. */
  #screenArguments(tool: unknown, args: Record<string, unknown>): void {
    const findings: string[] = [];

    const walk = (node: unknown, depth: number): void => {
      if (depth > MAX_ARGUMENT_DEPTH) {
        findings.push('nesting');
        return;
      }
      if (Array.isArray(node)) {
        for (const item of node) walk(item, depth + 1);
        return;
      }
      if (node !== null && typeof node === 'object') {
        for (const [key, value] of Object.entries(node as Record<string, unknown>)) {
          if (!ALLOWED_ARGUMENT_KEYS.has(key) && SECRET_SHAPED_FIELD.test(key)) {
            findings.push(`field:${safeDisplay(key, 40)}`);
          }
          walk(value, depth + 1);
        }
        return;
      }
      if (typeof node === 'string' && looksLikeCredential(node)) {
        findings.push('credential-shaped-value');
      }
    };

    walk(args, 0);
    if (findings.length > 0) {
      this.log.security('secret_shaped_argument_rejected', {
        tool: safeDisplay(String(tool), 40),
        detail: { findings: [...new Set(findings)].slice(0, 5) },
      });
      throw veilError(
        ErrorCode.FORBIDDEN_FIELD,
        'This tool never accepts credential material. Remove the offending field and let ' +
          'the user enter the value in Veil.',
      );
    }
  }

  async #store(args: Record<string, unknown>): Promise<ToolResult> {
    const params = parseStoreParams(args, this.registry);
    const request = await this.broker.createRequest(params);
    return { payload: this.#pendingPayload(request.requestId), isError: false };
  }

  async #revise(args: Record<string, unknown>): Promise<ToolResult> {
    const requestId = args.request_id;
    if (typeof requestId !== 'string') {
      throw veilError(ErrorCode.INVALID_ARGUMENTS, 'A request_id is required.');
    }
    const rest = Object.fromEntries(Object.entries(args).filter(([key]) => key !== 'request_id'));
    const params = parseStoreParams(rest, this.registry);
    const request = await this.broker.revise(requestId, params);
    const payload = this.#pendingPayload(request.requestId);
    payload.previous_request_id = safeDisplay(requestId, 64);
    payload.notice =
      'The previous authorization was invalidated. The user authorizes this operation again.';
    return { payload, isError: false };
  }

  async #status(args: Record<string, unknown>): Promise<ToolResult> {
    const parsed = statusArgsSchema.safeParse(args);
    if (!parsed.success) {
      throw veilError(ErrorCode.INVALID_ARGUMENTS, 'A request_id is required.');
    }
    const wait = parsed.data.wait_seconds ?? 0;
    if (wait > 0) {
      return {
        payload: await this.broker.waitForTerminal(parsed.data.request_id, wait),
        isError: false,
      };
    }
    return { payload: this.broker.publicStatus(parsed.data.request_id), isError: false };
  }

  #cancel(args: Record<string, unknown>): ToolResult {
    const parsed = cancelArgsSchema.safeParse(args);
    if (!parsed.success) {
      throw veilError(ErrorCode.INVALID_ARGUMENTS, 'A request_id is required.');
    }
    this.broker.cancel(parsed.data.request_id, {
      reason: parsed.data.reason ? safeDisplay(parsed.data.reason, 80) : 'agent_cancelled',
    });
    return { payload: this.broker.publicStatus(parsed.data.request_id), isError: false };
  }

  #destinations(): ToolResult {
    return {
      payload: {
        destinations: this.registry.list().map((adapter) => ({
          id: adapter.id,
          title: adapter.displayName,
          destination_class: adapter.destinationClass,
          write_modes: [...adapter.supportedWriteModes()],
          target_schema: adapter.targetSchema(),
        })),
        note:
          'No destination accepts a credential value through this API. Veil collects the ' +
          'value from the user directly.',
      },
      isError: false,
    };
  }

  #pendingPayload(requestId: string): Record<string, unknown> {
    return this.broker.publicStatus(requestId);
  }
}

function errorResult(error: VeilError): ToolResult {
  return { payload: publicErrorToJSON(error.public), isError: true };
}
